src/Build/Evaluation/LazyItemEvaluator.cs

```diff
@@ -219,17 +219,299 @@ public void MarkAsReferenced()
+                                    }
+                                    keysToRemove ??= new List<string>();
+                                    keysToRemove.Add(kvp.Key);
+                                }
+                            }
+
+                            if (keysToRemove != null)
+                            {
+                                foreach (string key in keysToRemove)
+                                {
+                                    dictionaryBuilder.Remove(key);
+                                }
+                            }
+                        }
+                    }
+
+                    // Finish by removing items from the list.
+                    if (items != null)
+                    {
+                        _listBuilder.RemoveAll(item => items.Contains(item.Item));
+                    }
+                }
+
+                /// <summary>
+                /// Creates an immutable view of this collection.
+                /// </summary>
+                public OrderedItemDataCollection ToImmutable()
+                {
+                    return new OrderedItemDataCollection(_version, _listBuilder.ToImmutable(), _dictionaryBuilder?.ToImmutable());
+                }
+
+                private IDictionary<string, DictionaryValue> GetOrCreateDictionaryBuilder()
+                {
+                    if (_dictionaryBuilder == null)
+                    {
+                        _dictionaryBuilder = ImmutableDictionary.CreateBuilder<string, DictionaryValue>(StringComparer.OrdinalIgnoreCase);
+                        foreach (ItemData item in _listBuilder)
+                        {
+                            AddToDictionary(item.Item);
+                        }
+                    }
+                    return _dictionaryBuilder;
+                }
+
+                private void AddToDictionary(I item)
+                {
+                    string key = FileUtilities.NormalizePathForComparisonNoThrow(item.EvaluatedInclude, item.ProjectDirectory);
+
+                    if (!_dictionaryBuilder.TryGetValue(key, out var dictionaryValue))
+                    {
+                        dictionaryValue = new DictionaryValue(_version, item);
+                    }
+                    dictionaryValue.Add(_version, item);
+                    _dictionaryBuilder[key] = dictionaryValue;
+                }
+            }
+
+            #endregion
+
+            /// <summary>
+            /// The current version of the collection.
+            /// </summary>
+            private int _version;
+
+            /// <summary>
+            /// The list of items in the collection. Defines the enumeration order.
+            /// </summary>
+            private ImmutableList<ItemData> _list;
+
+            /// <summary>
+            /// A dictionary of items keyed by their normalized value.
+            /// </summary>
+            private ImmutableDictionary<string, DictionaryValue> _dictionary;
```

Comment:
ImmutableDictionary is O(logn) indexing, this might be costly?

Reply:
In the grand scheme of things since N is ~number of files on disk, I believe that O(logN) is fine. The alternative would be to use a regular dictionary with O(1) access and rebuilding it every time immutability is actually exercised. Which happens when the collection is added to the cache in `MemoizedOperation`. There may be optimizations that would remove the immutability requirement but I don't want to go there in this PR.